### PR TITLE
WIP: Support for GHC 9.4

### PR DESCRIPTION
**Pull request by siraben**
## Description
Work in progress for GHC 9.4 support. Includes temporary overrides for optics and other packages in the flake. Testing is being performed on `aarch64-darwin`

## Checklist

- [ ] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

File: flake.lock [hunk 1]
```diff
@@ -155,6 +158,21 @@
         "rev": "1c8745c54a239d20b6fb0f79a8bd2628d779b27e",
         "type": "github"
       }
+    },
+    "systems": {
+      "locked": {
+        "lastModified": 1681028828,
+        "narHash": "sha256-Vy1rq5AaRuLzOxct8nz4T6wlgyUR7zLU309k9mBC768=",
+        "owner": "nix-systems",
```
Comment (msooseth) [inline flake.lock:166]: I don't really understand, is this necessary for this update to work?

**Review by msooseth — COMMENTED** [review]
Hi and thanks :) I just had a few questions, I thought I'd share! Let me know what you think,

Mate

File: flake.nix [hunk 2]
```diff
@@ -32,15 +32,46 @@
           z3
           cvc5
           git
-          foundry.defaultPackage.${system}
+          # foundry.defaultPackage.${system}
```
Comment (msooseth) [inline flake.nix:35]: Is there a reason why this needs to be removed?

Comment (siraben) [reply flake.nix]: Updated with conditional code. It failed on `aarch64-darwin`.

File: flake.lock (hunk 1 above)
Comment (siraben) [reply flake.lock:166]: This was added automatically by `nix flake update`.

Comment (msooseth) [reply flake.lock:166]: Thanks, makes sense

**Comment by siraben** [comment]
Right now the purpose is mainly to get a PR open so work isn't duplicated. Maybe I could separate out the `Bench.hs` fixes into a separate PR? Let me know!

**Comment by d-xo** [comment]
Thanks! Would be v happy to merge the bench.hs changes. Would be great if we would start building this in ci too.

File: bench/bench.hs [hunk 3]
```diff
@@ -84,25 +84,28 @@ runBCTest :: BCTests.Case -> IO Bool
...
  do
   let vm0 = BCTests.vmForCase x
-  result <- execStateT (Stepper.interpret (Fetch.zero 0 (Just 0)) . void $ Stepper.execFully) vm0
-  maybeReason <- BCTests.checkExpectation False x result
-  pure $ isNothing maybeReason
+  result <- Stepper.interpret (Fetch.zero 0 Nothing) vm0 Stepper.execFully
+  case result of
+    Left _ -> pure False
+    Right _ -> pure True
+  -- maybeReason <- BCTests.checkExpectation False x (_ result)
```
Comment (arcz) [inline bench/bench.hs:91]: is this no longer needed? can we remove commented lines?

File: flake.nix [hunk 4]
```diff
@@ -32,15 +32,47 @@
...
+              sha256 = "sha256-wzZ/64G7nVAzIFjKNs/jBvv6gQdTIQS4X/OvM5KWfnU=";
+            } + "/optics") {});
+            indexed-profunctors = pkgs.haskell.lib.dontCheck (self.callCabal2nix "optics" (pkgs.fetchFromGitHub {
+              owner = "well-typed";
+              repo = "optics";
+              rev = "46b03019bd7d9eddb767b19a68b94125eec3b17a";
+              sha256 = "sha256-wzZ/64G7nVAzIFjKNs/jBvv6gQdTIQS4X/OvM5KWfnU=";
+            } + "/indexed-profunctors") {});
+            # use obsidian systems fork of string-qq
+            string-qq = self.callCabal2nix "string-qq" (pkgs.fetchFromGitHub {
```
Comment (arcz) [inline flake.nix:65]: can we upstream this to nixpkgs?

File: flake.nix [hunk 5]
```diff
@@ -32,15 +32,47 @@
...
         ];
 
         secp256k1-static = stripDylib (pkgs.secp256k1.overrideAttrs (attrs: {
           configureFlags = attrs.configureFlags ++ [ "--enable-static" ];
         }));
 
+        patchedHaskellPackages = pkgs.haskell.packages.ghc94.override {
+          overrides = self: super: {
+            # disable tests in optics
+            optics-core = pkgs.haskell.lib.dontCheck (self.callCabal2nix "optics-core" (pkgs.fetchFromGitHub {
```
Comment (arcz) [inline flake.nix:46]: This will be included in the next bump right?

**Review by arcz — COMMENTED** [review]
I think we can have the bench fixes in a separate PR. The dependency changes will likely have to wait a bit because we need to first fix the tests on the newer version, see https://github.com/ethereum/hevm/pull/248.

File: bench/bench.hs (hunk 3 above)
Comment (siraben) [reply bench/bench.hs:91]: I was not sure if this is a good replacement. @d-xo what do you think?

File: flake.nix (hunk 5 above)
Comment (siraben) [reply flake.nix:46]: Yes, when the next-next `haskell-updates` is merged after https://github.com/NixOS/nixpkgs/pull/236431

File: flake.nix (hunk 4 above)
Comment (siraben) [reply flake.nix:65]: Sure, I'll make a PR to Nixpkgs explaining the situation with string-qq

File: bench/bench.hs (hunk 3 above)
Comment (d-xo) [reply bench/bench.hs:91]: I don't think this is a good replacement. `checkExpectation` is comparing various parts of the state to the expected results defined in the blockchain tests, this new code is just checking if the execution was succesful or not....

**Comment by d-xo** [comment]
Can this be closed now we have https://github.com/ethereum/hevm/pull/303 ?

**Comment by siraben** [comment]
Closing.

**Pull request closed without merging**